enhancements/rhcos/rhcos-inject.md

```diff
@@ -0,0 +1,198 @@
+---
+title: rhcos-inject
+authors:
+  - "@crawford"
+reviewers:
+  - TBD
+approvers:
+  - TBD
+creation-date: 2020-10-01
+last-updated: 2020-10-01
+status: provisional
+---
+
+# rhcos-inject
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+Throughout it's existence, OpenShift 4 has put a heavy influence on installation flows that make use of installer-provisioned infrastructure. This has largely been successful for predictable environments, such as AWS, Azure, or GCP, but it has (unsurprisingly) proven to make deployments into less predictable environments more difficult. These less predictable environments introduce a lot of variability into areas like network configuration, disk layout, and the life cycle of the machines and that makes it difficult or impossible for OpenShift to start from a foundation of shared assumption. To bridge this gap, admins need a way to inject a certain amount of customization before OpenShift installation can begin.
+
+## Motivation
+
+Admins need a way to inject customization into their RHCOS nodes before they are provisioned by the cluster. In most cases, this configuration is require in order for the provisioning process to complete, so the normal facilities (e.g. Machine Configuration Operator) are not yet available. Today, this is a very manual process involving an admin interactively providing that configuration at the point of running the RHCOS installer. This might work in some environments, but many others don't have easy interactive access due to technical or policy constraints.
+
+### Goals
+
+- Make it easy for an admin to add network configuration to each of the RHCOS nodes added both before and after OpenShift installation.
+- Allow an admin to add custom services and miscellaneous configuration to their RHCOS nodes.
+
+### Non-Goals
+
+- Inventory management
+- Bring-your-own-RHEL customization
+
+## Proposal
+
+Add a new, optional step to the installation process that allows an admin to inject network configuration and any other customization to the RHCOS installer image (and maybe other RHCOS images in the future) before invocation. This customization is performed by a new `rhcos-inject` utility that takes as input an RHCOS installer image (`rhcos-<version>-installer.<architecture>.iso`), network configuration, and a set of Ignition Configs and generates a new RHCOS installer image with all of the configuration and customization included. This new installer image can then be booted in an unsupervised manner and will complete the RHCOS installation along with any customization.
```

Comment:
It sounds as though this tool does exactly the following:
- Generate an Ignition config for the live system that embeds NM configuration + the Ignition config for the installed system + a `coreos-installer` unit that forwards them to the installed system with `--ignition-file` and `--copy-network` + any custom Ignition fragments
- Embed it with `coreos-installer iso ignition embed`

Is that the idea?  If so, it'd be good to spell out why this should be a new tool rather than a subcommand of coreos-installer.  If we view this as sugar over `coreos-installer iso ignition embed`, I don't think the functionality would be too far outside coreos-installer's existing scope, and implementing it as a subcommand would save us having to distribute an additional binary.

Also worth noting that some of the customization functionality would be useful for PXE installs, so we should also be able to produce the live Ignition config without embedding it into an ISO.

Nit: `rhcos-<version>-installer.<architecture>.iso` is the old installer image; I assume you mean `rhcos-<version>-live.<architecture>.iso`.

Reply:
I wasn't aware that the new coreos-installer had so much functionality. It sounds like a lot of the heavy-lifting has already been done and that these new features should just be folded in. That's great news!

Regarding the original question, I had assumed that the network config would happen for two contexts: the pre-pivot Ignition environment (initramfs) and the post-pivot system. This would allow for image-based deployments (e.g. QCOWs or VHDs) to fetch their Ignition Configs from the Machine Config Server running in the cluster. If we were only to configure the networking in the post-pivot system (e.g. by using Ignition to write the configs), the machine wouldn't be able to fetch its config from the MCS and would have to be provided the fully-rendered config up front. This is only easy to do if the live ISO is performing the installation though.